File: surf-cloud-api/src/main/java/dev/slne/surf/cloud/api/config/properties/SystemPropertyImpl.java
package dev.slne.surf.cloud.api.config.properties;

import java.util.function.Function;
import lombok.RequiredArgsConstructor;
import lombok.experimental.UtilityClass;
import org.jetbrains.annotations.Nullable;

@UtilityClass
class SystemPropertyImpl {

  <T> PropertyImpl<T> property(
      String key,
      Function<String, T> parser,
      @Nullable T defaultValue
  ) {
    return new PropertyImpl<>(key, parser, defaultValue);
  }

  @RequiredArgsConstructor
  private static final class PropertyImpl<T> implements SystemProperty<T> {

    private final String key;
    private final Function<String, T> parser;
    private final T defaultValue;
    private @Nullable T value;
    private boolean initialized;

    @Override
    public T value() {
      if (!initialized) {
        final String property = System.getProperty(key);
        this.value = property != null ? parser.apply(property) : defaultValue;
        this.initialized = true;
      }

      return this.value;
    }
  }
}

File: surf-cloud-api/src/main/kotlin/dev/slne/surf/cloud/api/config/properties/SystemPropertyImpl.kt
package dev.slne.surf.cloud.api.config.properties

class SystemPropertyImpl<T>(
    private val key: String,
    private val parser: (String) -> T,
    private val defaultValue: T? = null,
    private var value: T? = null,
    private var initialized: Boolean = false,
) : SystemProperty<T> {


    override fun value(): T? {
        if (!initialized) {
            val property = System.getProperty(key)
            this.value = if (property != null) parser(property) else defaultValue
            this.initialized = true
        }

        return this.value
    }
}
